File: 2024/src/03/day03b.js
const fs = require("fs");

fs.readFile("input/day03.in", "utf8", (err, data) => {
  let result = 0;
  let readInstruction = true;
  const pattern = /mul\((\d+),(\d+)\)|do\(\)|don't\(\)/g;
  const lines = data.split("\n");
  lines.forEach((line) => {
    const matches = line.matchAll(pattern);
    for (const match of matches) {
      if (match[0] === "do()") {
        readInstruction = true;
      } else if (match[0] === "don't()") {
        readInstruction = false;
      } else if (readInstruction) {
        result += Number(match[1]) * Number(match[2]);
      }
    }
  });
  console.log(result);
});

File: 2024/src/03/day03b.py
import re

pattern = re.compile(r"mul\((\d+),(\d+)\)|do\(\)|don't\(\)")

result = 0
readInstruction = True
with open("input/day03.in", "r") as file:
    for line in file:
        for match in pattern.finditer(line):
            if match[0] == "don't()":
                readInstruction = False
            elif match[0] == "do()":
                readInstruction = True
            else:
                if readInstruction:
                    result += int(match[1]) * int(match[2])
print(result)
